httpclient5-testing/src/test/java/org/apache/hc/client5/testing/compatibility/async/HttpAsyncClientCompatibilityTest.java

```diff
@@ -0,0 +1,174 @@
+
+import org.apache.hc.client5.http.ContextBuilder;
+import org.apache.hc.client5.http.async.methods.SimpleHttpRequest;
+import org.apache.hc.client5.http.async.methods.SimpleHttpResponse;
+import org.apache.hc.client5.http.async.methods.SimpleRequestBuilder;
+import org.apache.hc.client5.http.auth.AuthScope;
+import org.apache.hc.client5.http.auth.Credentials;
+import org.apache.hc.client5.http.auth.UsernamePasswordCredentials;
+import org.apache.hc.client5.http.impl.async.CloseableHttpAsyncClient;
+import org.apache.hc.client5.http.impl.auth.BasicCredentialsProvider;
+import org.apache.hc.client5.http.protocol.HttpClientContext;
+import org.apache.hc.client5.testing.extension.async.HttpAsyncClientResource;
+import org.apache.hc.core5.http.HttpHost;
+import org.apache.hc.core5.http.HttpStatus;
+import org.apache.hc.core5.http.HttpVersion;
+import org.apache.hc.core5.http2.HttpVersionPolicy;
+import org.apache.hc.core5.util.Timeout;
+import org.junit.jupiter.api.Assertions;
+import org.junit.jupiter.api.Test;
+import org.junit.jupiter.api.extension.RegisterExtension;
+
+public abstract class HttpAsyncClientCompatibilityTest {
+
+    static final Timeout TIMEOUT = Timeout.ofSeconds(5);
+
+    private final HttpVersionPolicy versionPolicy;
+    private final HttpHost target;
+    @RegisterExtension
+    private final HttpAsyncClientResource clientResource;
+    private final BasicCredentialsProvider credentialsProvider;
+
+    public HttpAsyncClientCompatibilityTest(
+            final HttpVersionPolicy versionPolicy,
+            final HttpHost target,
+            final HttpHost proxy,
+            final Credentials proxyCreds) throws Exception {
+        this.versionPolicy = versionPolicy;
+        this.target = target;
+        this.clientResource = new HttpAsyncClientResource(versionPolicy);
+        this.clientResource.configure(builder -> builder.setProxy(proxy));
+        this.credentialsProvider = new BasicCredentialsProvider();
+        if (proxy != null && proxyCreds != null) {
+            this.credentialsProvider.setCredentials(new AuthScope(proxy), proxyCreds);
+        }
+    }
+
+    CloseableHttpAsyncClient client() {
+        return clientResource.client();
+    }
+
+    HttpClientContext context() {
+        return ContextBuilder.create()
+                .useCredentialsProvider(credentialsProvider)
+                .build();
+    }
+
+    void addCredentials(final AuthScope authScope, final Credentials credentials) {
+        credentialsProvider.setCredentials(authScope, credentials);
+    }
+
+    void assertProtocolVersion(final HttpClientContext context) {
+        switch (versionPolicy) {
+            case FORCE_HTTP_1:
+                Assertions.assertEquals(HttpVersion.HTTP_1_1, context.getProtocolVersion());
+                break;
+            case FORCE_HTTP_2:
+            case NEGOTIATE:
+                Assertions.assertEquals(HttpVersion.HTTP_2, context.getProtocolVersion());
+                break;
```

Comment:
Do we want a `default` case that throws an exception in case something unexpected happens?